source_middleware.go

```diff
@@ -34,3 +59,350 @@ func SourceWithMiddleware(d Source, middleware ...SourceMiddleware) Source {
+			Description: "Whether to extract and encode the record key with a schema.",
+		},
+		configSourceSchemaKeySubject: {
+			Default: func() string {
+				if s.KeySubject != nil {
+					return *s.KeySubject
+				}
+				return ""
+			}(),
+			Type: config.ParameterTypeString,
+			Description: func() string {
+				desc := "The subject of the payload schema."
+				if s.KeySubject == nil {
+					desc += ` Defaults to the connector ID with a ".key" postfix.`
+				}
+				return desc
+			}(),
+		},
+	}
+}
+
+func (s SourceWithSchemaConfig) types() []string {
+	out := make([]string, 0, len(schema.KnownSerdeFactories))
+	for t := range schema.KnownSerdeFactories {
+		out = append(out, t.String())
+	}
+	return out
+}
+
+// SourceWithSchema is a middleware that extracts and encodes the record payload
+// and key with a schema. The schema is extracted from the record data for each
+// record produced by the source. The schema is registered with the schema
+// service and the schema subject is attached to the record metadata.
+type SourceWithSchema struct {
+	Config SourceWithSchemaConfig
+}
+
+// Wrap a Source into the schema middleware.
+func (s *SourceWithSchema) Wrap(impl Source) Source {
+	if s.Config.SchemaType == 0 {
+		s.Config.SchemaType = schema.TypeAvro
+	}
+
+	if s.Config.KeyEncode == nil {
+		t := true
+		s.Config.KeyEncode = &t
+	}
+	if s.Config.PayloadEncode == nil {
+		t := true
+		s.Config.PayloadEncode = &t
+	}
+
+	return &sourceWithSchema{
+		Source:           impl,
+		config:           s.Config,
+		fingerprintCache: make(map[uint64]schema.Schema),
+	}
+}
+
+// sourceWithSchema is the actual middleware implementation.
+type sourceWithSchema struct {
+	Source
+	config SourceWithSchemaConfig
+
+	schemaType     schema.Type
+	payloadSubject string
+	keySubject     string
+
+	fingerprintCache map[uint64]schema.Schema
+	payloadWarnOnce  sync.Once
+	keyWarnOnce      sync.Once
+}
+
+func (s *sourceWithSchema) Parameters() config.Parameters {
+	return mergeParameters(s.Source.Parameters(), s.config.parameters())
+}
+
+func (s *sourceWithSchema) Configure(ctx context.Context, config config.Config) error {
+	err := s.Source.Configure(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	connectorID := internal.ConnectorIDFromContext(ctx)
+
+	s.schemaType = s.config.SchemaType
+	if val, ok := config[configSourceSchemaType]; ok {
+		if err := s.schemaType.UnmarshalText([]byte(val)); err != nil {
+			return fmt.Errorf("invalid %s: failed to parse schema type: %w", configSourceSchemaType, err)
+		}
+	}
+
+	encodeKey := *s.config.KeyEncode
+	if val, ok := config[configSourceSchemaKeyEncode]; ok {
+		encodeKey, err = strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("invalid %s: failed to parse boolean: %w", configSourceSchemaKeyEncode, err)
+		}
+	}
+	if encodeKey {
+		// TODO: when adding schema context support, set DefaultKeySubject
+		//  to "key" and let the schema service attach the prefix / context.
+		s.keySubject = connectorID + ".key"
```

Comment:
We're going to this when handling name conflicts, no?

Reply:
Yeah, at that point this parameter can get a simple default value `"key"` and we will let the other parameter choose if a connector ID (or another prefix) gets added to make it unique and prevent name conflicts.

Reply:
So maybe we can just remove the connector ID from this code right away?

Reply:
But then this code won't behave as it should 😕

Reply:
Right... I can then change this in the name conflicts branches once everything is merged.